### DISCUSSION TITLE
partial_fit vs. fit

### DISCUSSION THREAD

**Discussion by noahberhe** [MaartenGr/BERTopic#1404] [Q&A]
Hi Maarten

What's the difference between using unsupervised .fit() on a large dataset, and then .transform() on incremental new data... vs. .partial_fit() on large dataset and then .partial_fit() on incremental new data? Unless we use River, am I right in saying there's no distinction as .partial_fit() doesn't discover new topics in incremental new data and tries to fit the new data to the pre-existing 50 clusters, and that the only distinction is the name given to those topics thanks to the decay parameter in the OnlineCountVectorizer, and also the fact that partial_fit() seems to fit all data to a cluster (so it has reduce_outliers() under the hood already?)?

Thanks,
Noah

**Comment by MaartenGr** [comment]
`partial_fit` and `fit` are actually quite different procedures and differ in a number of ways:

First, they both differ in their underlying models. Fit uses static models and calculates everything at once with UMAP and HDBSCAN, which is not possible with partial fit which needs IncrementalPCA and MiniBatch k-Means. Thus, partial fit needs models that allow for incremental learning. 

Second, when representing the topics the fit function has access to the full representation of documents and can often generate more accurate representations whereas partial fit needs to continuously update the underlying vectorizer to mimic the same behavior. During training, this can create poorer representations but should be more accurate as it is continuous training. 

Third, the generation of clusters generally differs because of incremental learning. Fit does this "right" at once whilst partial fit needs some time to accurately assign documents to clusters. As a result, this might mean that the assignment of earlier documents is poor compared to later documents. You can account for this by using transform after having run all data through partial fit and re-assigning the documents.

There might be more nuances to this but those are the most important ones I believe.

**Reply by noahberhe** [reply]
Thank you Maarten,
So it sounds like you'd normally only use .partial_fit() if you don't have enough data collected at the present moment, so you want the model to keep evolving and learning, or to guard against concept drift.

In my case I've got most of the data I need and drift is minimal, and I'm happy to proceed with the model to .transform() future data. However there is the possibility that new topics emerge (e.g. something new / niche happens next month that isn't in the dataset which needs to be captured in a new topic), so this means I need to use River? The only other option would be to retrain an Offline model in a month's time?

**Reply by MaartenGr** [reply]
Generally, you would use `.partial_fit` for two reasons. First, it is often used when you lack computational power and cannot load in the full dataset or train them. You would instead use `.partial_fit` to iteratively train the model such that you can prevent memory spikes. Second, it is used to for continuous training as new topics might appear in new data. This is often helpful in the case of support where we expect new topics/issues to arise as time goes on which the model would not have seen before. 

In your case, if discovering new topics is not the main goal of your solution, then it might be worthwhile to simply look for -1 topics and cluster those and inspect those.

**Reply by noahberhe** [reply]
> In your case, if discovering new topics is not the main goal of your solution, then it might be worthwhile to simply look for -1 topics and cluster those and inspect those.

That's interesting, so I can potentially compare these clusters to the clusters given by .reduce_outliers() from the existing model, and perhaps run some sort of similarity measure between the cluster labels? Could that be the best way to check if a new topic hasn't emerged in the evolving data, or is there a better way? Thank you !

**Reply by MaartenGr** [reply]
There are many ways in which you can check if a new topic has emerged. Generally, I would advise simply looking at the distribution of -1 documents, if they tend to increase over time, then that might mean that a new topic is slowly emerging.

**Reply by noahberhe** [reply]
Yes, that's what I was thinking to do also.
Thank you Maarten :)
